README.md
# :warning: DEPRECATION

Grafana Boom Theme Panel was created during early Grafana days to workaround theme support for Grafana dashbaords. Given that Grafana have native themes now, It is time to deprecate the plugin.

Thank you community for the support. This plugin will be removed/archived from the grafana plugin catalog. Existing installations should continue to work. If that doesn't work, reach out to [community.grafana.com](https://community.grafana.com) for support. You can also download old binaries from [releases page](https://github.com/yesoreyeram/yesoreyeram-boomtheme-panel/releases).

# Boom Theme Panel

[![Build & Publish](https://github.com/yesoreyeram/yesoreyeram-boomtheme-panel/workflows/Build%20&%20Publish/badge.svg)](https://github.com/yesoreyeram/yesoreyeram-boomtheme-panel/actions?query=workflow%3A%22Build+%26+Publish%22)
[![Release](https://github.com/yesoreyeram/yesoreyeram-boomtheme-panel/workflows/Release/badge.svg)](https://github.com/yesoreyeram/yesoreyeram-boomtheme-panel/actions?query=workflow%3ARelease)

Theme switcher with custom styles / themes for grafana dashboards.

![image](https://user-images.githubusercontent.com/153843/99880461-5e1aa900-2c0b-11eb-9c28-3c80381f2504.png)
![image](https://user-images.githubusercontent.com/153843/99880502-a20dae00-2c0b-11eb-8355-c7c87bcb2f5c.png)

## Features

- Theme switcher
- Multiple themes per dashboard
- Theme builder
- External stylesheets support
- Add inline styles to themes to override styles
- Add many themes as possible without rebuilding / restarting grafana
- Users can change the themes without editing the dashboard
- Dashboard specific themes

## Creating Theme

Themes can be created with multiple building blocks like background image, base theme etc.

| Property                    | Description                                                                                              |
| --------------------------- | -------------------------------------------------------------------------------------------------------- |
| Base theme                  | Themes can be built on top of default/dark/light theme. Default is **Default Theme**                     |
| Background image            | Optional property. Can be blank. If specified more than once, last wins. Value should be valid image URL |
| CSS url                     | External theme file. Should be valid CSS file URL                                                        |
| Custom Style / CSS Override | CSS Styles. Should be valid css                                                                          |
| Panel container BG          | Background color for the panels                                                                          |

## Supported Grafana version

This grafana plugin is tested with the following grafana versions, But other versions are also expected to work.

- Grafana version 8.x
- Grafana version 7.x
- Grafana version 6.x

## Notes

- When adding external stylesheets, make sure CORS enabled for those domains.

- To make panel invisible : Modify following theme panel settings:
  - transparent = true
  - title = ""
  - Disable Theme Picker using panel settings
  - Move this panel to the bottom of the dashboard
  - Adjust the height and width if required.

## Known issues / Limitations

- If any custom plugin is used, dark/light theme switch, base theme will not work for those custom plugins. Refer [this](https://github.com/yesoreyeram/yesoreyeram-boomtheme-panel/issues/3) github issue

- Theme panel should be within view port. Otherwise, Grafana won't render the theme. Refer [this](https://github.com/yesoreyeram/yesoreyeram-boomtheme-panel/issues/17)

## Installation Instructions

There are multiple ways to install this plugin into your grafana instance

### Download and extract zip file

Download the zip file from [github releases page](https://github.com/yesoreyeram/yesoreyeram-boomtheme-panel/releases) and extract into your grafana plugin folder. Then restart Grafana.

### Using grafana-cli

If you are using grafana-cli, execute the following command to install the plugin

```sh
grafana-cli plugins install yesoreyeram-boomtheme-panel
```

or for specific versions

```sh
grafana-cli --pluginUrl https://github.com/yesoreyeram/yesoreyeram-boomtheme-panel/releases/download/v0.2.1/yesoreyeram-boomtheme-panel-0.2.1.zip plugins install yesoreyeram-boomtheme-panel
```

### Using helm chart

If you use helm chart to provision grafana, use the following config to install the plugin

```yml
plugins:
  - yesoreyeram-boomtheme-panel
```

or for any specific versions

```yml
plugins:
  - https://github.com/yesoreyeram/yesoreyeram-boomtheme-panel/releases/download/v0.2.1/yesoreyeram-boomtheme-panel-0.2.1.zip;yesoreyeram-boomtheme-panel
```


CHANGELOG.md
# CHANGELOG

| Version | Changes                                            |
| ------- | -------------------------------------------------- |
| 0.3.0   | Plugin deprecated                                  |
| 0.2.1   | Bug fixes (Thanks @koszti)                         |
|         | Replace theme by variable (Thanks @koszti)         |
| 0.2.0   | **Plugin Signed**                                  |
|         | Tagged releases                                    |
|         | Bug fixes                                          |
| 0.1.0   | **First version for release**                      |
| 0.0.4   | Base theme option                                  |
|         | Bug fixes (#4)                                     |
| 0.0.3   | Schema changes (BREAKING)                          |
|         | Option to switch between default dark,light themes |
|         | BG Image as theme option                           |
|         | Bug fixes                                          |
| 0.0.2   | Multiple Themes, Theme picker                      |
| 0.0.1   | Base version                                       |


tests/default.spec.ts
import _ from "lodash";
describe("Default Test Suite", () => {
  it("Default Test 1", () => {
    expect("Hello").toBe("Hello");
  });
  it("Default Test 2", () => {
    expect(1).toBe(1);
  });
});


GruntFile.js
const sass = require("node-sass");

module.exports = grunt => {
  require("load-grunt-tasks")(grunt);

  grunt.initConfig({
    clean: ["dist"],

    copy: {
      src_to_dist: {
        cwd: "src",
        expand: true,
        src: ["**/*", "!**/*.ts", "!**/*.js", "!**/*.scss", "!img/**/*"],
        dest: "dist"
      },
      pluginDef: {
        expand: true,
        src: ["plugin.json", "*.md"],
        dest: "dist"
      },
      img_to_dist: {
        cwd: "src",
        expand: true,
        src: ["img/**/*"],
        dest: "dist/src/"
      },
      logo_to_dist: {
        cwd: "src",
        expand: true,
        src: ["img/logo.png"],
        dest: "dist/"
      }
    },

    watch: {
      rebuild_all: {
        files: ["src/**/*", "plugin.json", "README.md", "CHANGELOG.md"],
        tasks: [
          "ts:default",
          "sass:build",
          "copy:src_to_dist",
          "copy:pluginDef",
          "copy:img_to_dist"
        ],
        options: {
          spawn: false
        }
      }
    },

    sass: {
      build: {
        options: {
          debugInfo: true,
          check: true,
          implementation: sass,
          sourceMap: false
        },
        files: {
          "dist/css/default.dark.css": "src/scss/default.dark.scss",
          "dist/css/default.light.css": "src/scss/default.light.scss"
        }
      }
    },

    tslint: {
      options: {
        configuration: "tslint.json"
      },
      files: {
        src: ["src/**/*.ts"]
      }
    },

    ts: {
      default: {
        tsconfig: "./tsconfig.json"
      }
    },
  });

  grunt.registerTask("test", ["tslint"]);

  grunt.registerTask("default", [
    "clean",
    "tslint",
    "ts:default",
    "sass:build",
    "copy:src_to_dist",
    "copy:pluginDef",
    "copy:img_to_dist",
    "copy:logo_to_dist"
  ]);
};


src/utils.ts
import { BoomTheme } from './BoomTheme';

export const getThemeCSSFile = function (mode: 'dark' | 'light'): string {
  let fileName = '';
  if (['dark', 'light'].indexOf(mode.toLowerCase()) > -1 && window.performance) {
    let appFiles = window.performance
      .getEntries()
      .map((e) => e.name)
      .filter((e) => e.endsWith('.js'))
      .filter((e) => e.indexOf('/public/build/app.') > -1);
    if (appFiles && appFiles.length > 0) {
      fileName =
        appFiles[0].replace(`/public/build/app.`, `/public/build/grafana.${mode.toLowerCase()}.`).slice(0, -3) + '.css';
    }
  }
  return fileName;
};
export const getActiveThemeName = function (themes: BoomTheme[], index: number): string {
  switch (index) {
    case -1000:
      return 'Grafana Default';
    case -2000:
      return 'Grafana Dark';
    case -3000:
      return 'Grafana Light';
    default:
      return themes[index] && themes[index].name ? themes[index].name : 'Grafana Default';
  }
};


src/module.ts
import { PanelPlugin, PanelModel } from '@grafana/data';
import { Panel, PanelOptions } from './Panel';
import { ThemesEditorOptions } from './editors/ThemesEditor';
import { getActiveThemeName } from './utils';

export const plugin = new PanelPlugin<PanelOptions>(Panel)
  .setPanelOptions((builder) => {
    builder.addCustomEditor(ThemesEditorOptions);
    builder.addBooleanSwitch({
      name: 'Hide Theme Picker',
      path: 'disableThemePicker',
      category: ['Themes Options'],
    });
    builder.addTextInput({
      name: 'Default Theme',
      path: 'activeTheme',
      category: ['Themes Options'],
    });
    return builder;
  })
  .setMigrationHandler((panel: PanelModel<PanelOptions> & Record<string, any>) => {
    const newOptions = {
      disableThemePicker: panel.options.disableThemePicker ?? panel.disableThemePicker,
      activeTheme: panel.options.activeTheme ?? getActiveThemeName(panel.themes, panel.activeThemeId),
      themes: panel.options.themes ?? panel.themes,
    };
    delete panel.themes;
    delete panel.activeThemeId;
    delete panel.disableThemePicker;
    return newOptions;
  });


src/BoomTheme.ts
import { BoomThemeStyle, BoomThemeStyleProps } from './BoomThemeStyle';
import { getThemeCSSFile } from './utils';
import { CONFIG } from './config';

interface BoomThemeInterface {
  name: string;
  styles: BoomThemeStyle[];
}

export class BoomTheme {
  name: string;
  styles: BoomThemeStyle[];
  constructor(options: BoomThemeInterface) {
    this.name = options.name || CONFIG.DEFAULT_THEME_NAME;
    this.styles =
      options.styles.length === 0
        ? [
            new BoomThemeStyle(BoomThemeStyleProps.BaseTheme, { theme: CONFIG.BASE_THEMES.DEFAULT.id }),
            new BoomThemeStyle(BoomThemeStyleProps.BackgroundImage, { url: '' }),
            new BoomThemeStyle(BoomThemeStyleProps.ExternalURL, { url: '' }),
            new BoomThemeStyle(BoomThemeStyleProps.CustomStyle, { text: `` }),
          ]
        : options.styles;
  }
  addStyle(type: BoomThemeStyleProps): void {
    this.styles.push(new BoomThemeStyle(type, {}));
  }
  deleteStyle(index: number): void {
    this.styles.splice(index, 1);
  }
  private constructTheme(styles: any[] = []): string {
    let output = ``;
    styles.forEach((style) => {
      if (style.type === CONFIG.THEME_STYLES.URL) {
        if (style.props && style.props.url !== '') {
          output += `@import url('${style.props.url}');
                    `;
        }
      } else if (style.type === CONFIG.THEME_STYLES.BASE_THEME) {
        if (style.props && style.props.theme !== '') {
          if (style.props.theme.toLowerCase() === CONFIG.BASE_THEMES.DARK.id) {
            output += `@import url('${getThemeCSSFile('dark')}');
                        `;
          } else if (style.props.theme.toLowerCase() === CONFIG.BASE_THEMES.LIGHT.id) {
            output += `@import url('${getThemeCSSFile('light')}');
                        `;
          }
        }
      } else if (style.type === CONFIG.THEME_STYLES.STYLE) {
        if (style.props && style.props.text !== '') {
          output += `${style.props.text || ''}
                    `;
        }
      } else if (style.type === CONFIG.THEME_STYLES.BG_IMAGE) {
        if (style.props && style.props.url !== '') {
          output += `
.main-view, .sidemenu, .sidemenu-open .sidemenu, .navbar, .dashboard-container,.dashboard-container > div, .page-toolbar  {
    background: url("${style.props.url}") no-repeat center center fixed;
    -webkit-background-size: cover;
    -moz-background-size: cover;
    -o-background-size: cover;
    background-size: cover;
}
                    `;
        }
      } else if (style.type === CONFIG.THEME_STYLES.PANEL_CONTAINER_BG_COLOR) {
        if (style.props && style.props.color !== '') {
          output += `
.panel-container, .page-toolbar {
    background-color: ${style.props.color}
}
`;
        }
      }
    });
    return output;
  }
  getThemeContent(): string {
    let output = '';
    if (this.styles && this.styles.length > 0) {
      output += this.constructTheme(this.styles.filter((style) => style.type === CONFIG.THEME_STYLES.URL));
      output += this.constructTheme(this.styles.filter((style) => style.type === CONFIG.THEME_STYLES.STYLE));
      output += this.constructTheme(
        this.styles.filter(
          (style) => style.type !== CONFIG.THEME_STYLES.URL && style.type !== CONFIG.THEME_STYLES.STYLE
        )
      );
    }
    return output;
  }
}


src/BoomThemeCtl.ts
///<reference path="../node_modules/grafana-sdk-mocks/app/headers/common.d.ts" />

import _ from "lodash";
import { PanelCtrl } from "app/plugins/sdk";
import { BoomTheme } from './BoomTheme';
import { BoomThemeStyle } from './BoomThemeStyle';
import { getThemeCSSFile } from './utils';
import { CONFIG } from './config';

export class BoomThemeCtl extends PanelCtrl {
    public static templateUrl = "partials/module.html";
    public scope: any;
    public ctrl: any;
    public elem: any;
    public attrs: any;
    public activeEditorTabIndex: number;
    public runtimeThemeSet: Boolean;
    public runtimeThemeIndex: number;
    public base_theme_options: any = [CONFIG.BASE_THEMES.DEFAULT, CONFIG.BASE_THEMES.DARK, CONFIG.BASE_THEMES.LIGHT].map(theme => {
        return {
            text: theme.name,
            value: theme.id
        };
    });
    constructor($scope, $injector) {
        super($scope, $injector);
        _.defaults(this.panel, {});
        this.panel.themes = this.panel.themes || [
            new BoomTheme({
                name: CONFIG.FIRST_THEME_NAME,
                styles: [
                    new BoomThemeStyle(CONFIG.THEME_STYLES.BASE_THEME, { theme: CONFIG.BASE_THEMES.DEFAULT.id }),
                    new BoomThemeStyle(CONFIG.THEME_STYLES.BG_IMAGE, { url: CONFIG.DEFAULT_THEME_BG_IMAGE }),
                    new BoomThemeStyle(CONFIG.THEME_STYLES.STYLE, { text: CONFIG.DEFAULT_THEME_STYLE })
                ]
            })
        ];
        this.panel.activeThemeId = this.panel.activeThemeId || 0;
        this.activeEditorTabIndex = this.panel.activeThemeId >= 0 ? this.panel.activeThemeId : -1;
        this.runtimeThemeSet = false;
        this.runtimeThemeIndex = this.panel.activeThemeId >= 0 ? this.panel.activeThemeId : 0;
        this.updatePrototypes();
        this.events.on("init-edit-mode", this.onInitEditMode.bind(this));
    }
    private updatePrototypes(): void {
        if (this.panel.themes && this.panel.themes.length > 0) {
            this.panel.themes.map(theme => {
                Object.setPrototypeOf(theme, BoomTheme.prototype);
                if (theme.styles && theme.styles.length > 0) {
                    theme.styles.map(style => {
                        Object.setPrototypeOf(style, BoomThemeStyle.prototype);
                    });
                }
                return theme;
            });
        }
    }
    private onInitEditMode(): void {
        this.addEditorTab("Theme", "public/plugins/yesoreyeram-boomtheme-panel/partials/options.html", 2);
    }
    public addTheme(): void {
        this.panel.themes = this.panel.themes || [];
        this.panel.themes.push(new BoomTheme({
            name: `Theme ${this.panel.themes.length + 1}`
        }));
        this.activeEditorTabIndex = this.panel.themes.length - 1;
        this.runtimeThemeIndex = this.panel.themes.length - 1;
        this.render();
    }
    public deleteTheme(index: number): void {
        this.panel.themes.splice(index, 1);
        if (this.panel.activeThemeId === index) {
            this.panel.activeThemeId = 0;
        }
        if (this.runtimeThemeIndex === index) {
            this.runtimeThemeIndex = 0;
        }
        if (this.activeEditorTabIndex === index) {
            this.activeEditorTabIndex = this.ctrl.panel.themes.length - 1;
        }
    }
    public setThemeAsDefault(index: number) {
        this.ctrl.panel.activeThemeId = index;
        this.runtimeThemeIndex = index;
        this.render();
    }
    public setRunTimeThemeIndex(index: number) {
        this.runtimeThemeSet = true;
        this.runtimeThemeIndex = index;
        this.render();
    }
    public limitText(text: string, maxLength: Number): string {
        if (text.split("").length > maxLength) {
            text = text.substring(0, Number(maxLength) - 3) + "...";
        }
        return text;
    }
    public link(scope: any, elem: any, attrs: any, ctrl: any): void {
        this.scope = scope;
        this.elem = elem;
        this.attrs = attrs;
        this.ctrl = ctrl;
        this.render();
    }
}

BoomThemeCtl.prototype.render = function () {

    // #region Panel UI Options
    if (this.ctrl.panel.title === "Panel Title") {
        this.ctrl.panel.title = "";
    }
    if (this.ctrl.panel.gridPos && this.ctrl.panel.gridPos.x === 0 && this.ctrl.panel.gridPos.y === 0) {
        this.ctrl.panel.gridPos.w = 24;
        this.ctrl.panel.gridPos.h = 3;
    }
    // #endregion

    // #region Themes Rendering
    let output = '';
    _.each(this.panel.themes, (theme, index) => {
        if (this.runtimeThemeSet === false) {
            if (this.panel.activeThemeId === index && this.panel.activeThemeId >= 0) {
                output += theme.getThemeContent();
            }
        } else {
            if (this.runtimeThemeIndex === index && this.runtimeThemeIndex >= 0) {
                output += theme.getThemeContent();
            }
        }
    });
    if (this.runtimeThemeSet === true) {
        if (this.runtimeThemeIndex === CONFIG.BASE_THEMES.DARK.index_id) {
            output += `@import url('${getThemeCSSFile(CONFIG.BASE_THEMES.DARK.id)}');
            `;
        } else if (this.runtimeThemeIndex === CONFIG.BASE_THEMES.LIGHT.index_id) {
            output += `@import url('${getThemeCSSFile(CONFIG.BASE_THEMES.LIGHT.id)}');
            `;
        }
    }
    const style = document.createElement('style');
    style.type = 'text/css';
    style.appendChild(document.createTextNode(output));
    this.elem.find("#boom-theme").html("");
    this.elem.find("#boom-theme").append(style);
    // #endregion

};

export { BoomThemeCtl as PanelCtrl };


src/BoomThemeStyle.ts
import { CONFIG } from './config';
export enum BoomThemeStyleProps {
  BaseTheme = 'basetheme',
  BackgroundImage = 'bgimage',
  CustomStyle = 'style',
  ExternalURL = 'url',
  PanelBackground = 'panel-container-bg-color',
  None = 'none',
}
interface BoomThemeStyleInterface {
  type: BoomThemeStyleProps;
}
interface BoomThemeStyleBaseThemeInterface extends BoomThemeStyleInterface {
  type: BoomThemeStyleProps.BaseTheme;
  props: {
    theme: 'default' | 'dark' | 'light';
  };
}
interface BoomThemeStyleBgImageInterface extends BoomThemeStyleInterface {
  type: BoomThemeStyleProps.BackgroundImage;
  props: {
    url: string;
  };
}
interface BoomThemeStyleStyleInterface extends BoomThemeStyleInterface {
  type: BoomThemeStyleProps.CustomStyle;
  props: {
    text: string;
  };
}
interface BoomThemeStyleURLInterface extends BoomThemeStyleInterface {
  type: BoomThemeStyleProps.ExternalURL;
  props: {
    url: string;
  };
}
interface BoomThemePanelContainerBGColorInterface extends BoomThemeStyleInterface {
  type: BoomThemeStyleProps.PanelBackground;
  props: {
    color: string;
  };
}
export type BoomThemeStyleType =
  | BoomThemeStyleBaseThemeInterface
  | BoomThemeStyleBgImageInterface
  | BoomThemeStyleStyleInterface
  | BoomThemeStyleURLInterface
  | BoomThemePanelContainerBGColorInterface;

export class BoomThemeStyle implements BoomThemeStyleInterface {
  type: BoomThemeStyleProps;
  props: any;
  constructor(type: BoomThemeStyleProps, props: any) {
    this.type = type;
    switch (type.toLowerCase()) {
      case BoomThemeStyleProps.BaseTheme:
        this.props = {
          theme: props && props.theme ? props.theme : '',
        };
        break;
      case BoomThemeStyleProps.CustomStyle:
        this.props = {
          text: props && props.text ? props.text : '',
        };
        break;
      case BoomThemeStyleProps.ExternalURL:
        this.props = {
          url: props && props.url ? props.url : '',
        };
        break;
      case BoomThemeStyleProps.BackgroundImage:
        this.props = {
          url: props && props.url ? props.url : '',
        };
        break;
      case CONFIG.THEME_STYLES.PANEL_CONTAINER_BG_COLOR:
        this.props = {
          color: props && props.color ? props.color : '',
        };
        break;
      default:
        this.props = {};
        break;
    }
  }
}


src/config.ts
export const CONFIG = {
  BASE_THEMES: {
    DARK: {
      id: 'dark',
      index_id: -2000,
      name: 'Dark Theme',
    },
    DEFAULT: {
      id: 'default',
      index_id: -1000,
      name: 'Default Theme',
    },
    LIGHT: {
      id: 'light',
      index_id: -3000,
      name: 'Light Theme',
    },
  },
  DEFAULT_THEME_BG_IMAGE: `https://images.unsplash.com/photo-1524334228333-0f6db392f8a1`, // Image Credits : https://unsplash.com/photos/fX-qWsXl5x8
  DEFAULT_THEME_NAME: 'New Theme',
  DEFAULT_THEME_STYLE: `.panel-container {
    background-color: rgba(0,0,0,0.3);
}`,
  FIRST_THEME_NAME: 'Night Theme',
  THEME_STYLES: {
    BASE_THEME: 'basetheme',
    BG_IMAGE: 'bgimage',
    NONE: 'none',
    STYLE: 'style',
    PANEL_CONTAINER_BG_COLOR: 'panel-container-bg-color',
    URL: 'url',
  },
};
